In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

import os
# print(os.listdir("../input"))

# Any results you write to the current directory are saved as output.

In [30]:
df= pd.read_csv('C:/Users/rgangurde/Downloads/train.csv')

In [4]:
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [5]:
df.shape

(1460, 81)

In [6]:
X= df.drop('Id', axis= 1)

In [7]:
X= X.drop('SalePrice', axis= 1)

In [8]:
X.head(10)

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          60       RL         65.0     8450   Pave   NaN      Reg   
1          20       RL         80.0     9600   Pave   NaN      Reg   
2          60       RL         68.0    11250   Pave   NaN      IR1   
3          70       RL         60.0     9550   Pave   NaN      IR1   
4          60       RL         84.0    14260   Pave   NaN      IR1   
5          50       RL         85.0    14115   Pave   NaN      IR1   
6          20       RL         75.0    10084   Pave   NaN      Reg   
7          60       RL          NaN    10382   Pave   NaN      IR1   
8          50       RM         51.0     6120   Pave   NaN      Reg   
9         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities LotConfig  ... ScreenPorch PoolArea PoolQC  Fence  \
0         Lvl    AllPub    Inside  ...           0        0    NaN    NaN   
1         Lvl    AllPub       FR2  ...           0        0    NaN    NaN   
2         Lvl    AllPub    Inside  ...           0        0    NaN    NaN   
3         Lvl    AllPub    Corner  ...           0        0    NaN    NaN   
4         Lvl    AllPub       FR2  ...           0        0    NaN    NaN   
5         Lvl    AllPub    Inside  ...           0        0    NaN  MnPrv   
6         Lvl    AllPub    Inside  ...           0        0    NaN    NaN   
7         Lvl    AllPub    Corner  ...           0        0    NaN    NaN   
8         Lvl    AllPub    Inside  ...           0        0    NaN    NaN   
9         Lvl    AllPub    Corner  ...           0        0    NaN    NaN   

  MiscFeature MiscVal  MoSold  YrSold  SaleType  SaleCondition  
0         NaN       0       2    2008        WD         Normal  
1         NaN       0       5    2007        WD         Normal  
2         NaN       0       9    2008        WD         Normal  
3         NaN       0       2    2006        WD        Abnorml  
4         NaN       0      12    2008        WD         Normal  
5        Shed     700      10    2009        WD         Normal  
6         NaN       0       8    2007        WD         Normal  
7        Shed     350      11    2009        WD         Normal  
8         NaN       0       4    2008        WD        Abnorml  
9         NaN       0       1    2008        WD         Normal  

[10 rows x 79 columns]

In [9]:
X.dtypes

MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
Alley             object
LotShape          object
LandContour       object
Utilities         object
LotConfig         object
LandSlope         object
Neighborhood      object
Condition1        object
Condition2        object
BldgType          object
HouseStyle        object
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle         object
RoofMatl          object
Exterior1st       object
Exterior2nd       object
MasVnrType        object
MasVnrArea       float64
ExterQual         object
ExterCond         object
Foundation        object
BsmtQual          object
                  ...   
HalfBath           int64
BedroomAbvGr       int64
KitchenAbvGr       int64
KitchenQual       object
TotRmsAbvGrd       int64
Functional        object
Fireplaces         int64
FireplaceQu       object
GarageType        object


Creating an arbitrary model using Multiple Linear Regression Model

In [10]:
X_Features= X[['MSSubClass','LotFrontage','LotArea','OverallQual','OverallCond','YearBuilt','YearRemodAdd','MasVnrArea','HalfBath','BedroomAbvGr','KitchenAbvGr','TotRmsAbvGrd','Fireplaces','GarageYrBlt','GarageCars','GarageArea','WoodDeckSF','OpenPorchSF','EnclosedPorch','PoolArea','MiscVal','MoSold','YrSold']]

In [11]:
X_Features.fillna(X_Features)
X_Features= np.asanyarray(X_Features.values)
X_Features.shape

(1460, 23)

In [12]:
import math

for i in range(1460):
    for j in range(23):
        arr= X_Features[i][j]
        if math.isnan(arr) :
            X_Features[i][j]= 0

In [13]:
from sklearn.model_selection import train_test_split

Y= np.asanyarray(df['SalePrice'].values)
X_train, X_test, Y_train, Y_test= train_test_split(X_Features, Y, test_size= 0.3, random_state= 4)

In [14]:
from sklearn.linear_model import LinearRegression

model= LinearRegression()
model.fit(X_train, Y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [15]:
yHat= model.predict(X_test)

Evaluating the model

In [16]:
def rmsle(y_pred, y_test) : 
    assert len(y_test) == len(y_pred)
    return np.sqrt(np.mean((np.log(1+y_pred) - np.log(1+y_test))**2))

In [17]:
from sklearn import metrics

print('MSE:', (np.mean(Y_test - yHat)**2))
print('R2 score:', metrics.r2_score(yHat, Y_test))
print('RSMLE:',rmsle(np.absolute(yHat), np.absolute(Y_test)))

MSE: 1925234.4019549354
R2 score: 0.7960038877454458
RSMLE: 0.21417201106996459


With multiple linear regression, rsmle of 0.21417 was achieved.
The model consisted of int and float values only. So a vast amount of features were neglected for the training of model.
Looking ahead, more features will be analysed using different models

In [18]:
from sklearn.preprocessing import LabelEncoder

def categTOnumeric(features_):
    lb= LabelEncoder()
    for i in range(len(features_)):
        lb.fit(df_[features_[i]].astype('str'))
        df_[features_[i]]= lb.transform(df_[features_[i]].astype('str'))


## Creating a decision tree model using some of the labelled features 

In [19]:
categ= ['Street','Alley','LotShape','LandContour','Utilities','LotConfig','LandSlope','Neighborhood','Condition1','Condition2','BldgType','HouseStyle','ExterQual','ExterCond','Foundation','BsmtQual','GarageType','GarageCond','GarageQual','PavedDrive']
df_= df
categTOnumeric(categ)

In [20]:
categ= ['SaleType','SaleCondition']
categTOnumeric(categ)
nodes= df[['Street','Alley','LotShape','LandContour','Utilities','LotConfig','LandSlope','Neighborhood','Condition1','Condition2','BldgType','HouseStyle','ExterQual','ExterCond','Foundation','BsmtQual','GarageType','GarageCond','GarageQual','PavedDrive']].values

In [21]:
target= df[['SalePrice']].values

In [22]:
x_train, x_test, y_train, y_test= train_test_split(nodes, target, test_size= 0.3, random_state= 5)

In [23]:
from sklearn.tree import DecisionTreeClassifier

tree= DecisionTreeClassifier(criterion= 'entropy', max_depth= 20)
tree.fit(x_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=20,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [24]:
yHat_= tree.predict(x_test)
print('RSMLE:',rmsle(yHat_, y_test))

RSMLE: 0.5406596957920142


# Using K-Nearest Neighbors Classifier

In [25]:
from sklearn.neighbors import KNeighborsClassifier
neigh= KNeighborsClassifier(n_neighbors= 5).fit(x_train, y_train)

C:\Users\rgangurde\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  


In [26]:
yHat_Neigh= neigh.predict(x_test)

In [27]:
print('RSMLE:',rmsle(yHat_Neigh, y_test, ))

RSMLE: 0.5804533607579861


So far, multiple linear regression has been the most precise. So the focus will be on that model

In [409]:
df_.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street             int32
Alley              int32
LotShape           int32
LandContour        int32
Utilities          int32
LotConfig          int32
LandSlope          int32
Neighborhood       int32
Condition1         int32
Condition2         int32
BldgType           int32
HouseStyle         int32
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle          int32
RoofMatl           int32
Exterior1st        int32
Exterior2nd        int32
MasVnrType         int32
MasVnrArea       float64
ExterQual          int32
ExterCond          int32
Foundation         int32
                  ...   
BedroomAbvGr       int64
KitchenAbvGr       int64
KitchenQual        int32
TotRmsAbvGrd       int64
Functional         int32
Fireplaces         int64
FireplaceQu        int32
GarageType         int32
GarageYrBlt      float64


In [506]:
categTOnumeric(['MSZoning'])

In [507]:
X_Features_= df_[['YearBuilt','YearRemodAdd','LotFrontage','LotArea','OverallQual','OverallCond','MasVnrArea','BsmtFinSF1','BsmtFinSF2','BsmtUnfSF','TotalBsmtSF','1stFlrSF','2ndFlrSF','LowQualFinSF','GrLivArea','BsmtFullBath','BsmtHalfBath','FullBath','HalfBath','BedroomAbvGr','KitchenAbvGr','TotRmsAbvGrd','Fireplaces','GarageYrBlt','GarageCars','GarageArea','WoodDeckSF','OpenPorchSF','EnclosedPorch','PoolArea','YrSold','Street','Alley','LotShape','LandContour','Utilities','LotConfig','Condition2','BldgType','Exterior1st','MasVnrType','BsmtQual','BsmtCond','BsmtFinType2','Heating','HeatingQC','FireplaceQu','Fence','SaleCondition']].values

In [508]:
Y_Feature_= df['SalePrice'].values

In [500]:
X_Features_[0]

array([2.003e+03, 2.003e+03, 6.500e+01, 8.450e+03, 7.000e+00, 5.000e+00,
       1.960e+02, 7.060e+02, 0.000e+00, 1.500e+02, 8.560e+02, 8.560e+02,
       8.540e+02, 0.000e+00, 1.710e+03, 1.000e+00, 0.000e+00, 2.000e+00,
       1.000e+00, 3.000e+00, 1.000e+00, 8.000e+00, 0.000e+00, 2.003e+03,
       2.000e+00, 5.480e+02, 0.000e+00, 6.100e+01, 0.000e+00, 0.000e+00,
       2.008e+03, 1.000e+00, 2.000e+00, 3.000e+00, 3.000e+00, 0.000e+00,
       4.000e+00, 2.000e+00, 0.000e+00, 1.200e+01, 1.000e+00, 2.000e+00,
       3.000e+00, 5.000e+00, 1.000e+00, 0.000e+00, 5.000e+00, 4.000e+00,
       4.000e+00])

In [509]:
for i in range(len(X_Features_[:])):
    for j in range(len(X_Features_[0])):
        arr= X_Features_[i][j]
        if math.isnan(arr) :
            X_Features_[i][j]= 0

In [510]:
x_train, x_test, y_train, y_test= train_test_split(X_Features_, Y_Feature_, test_size= 0.3, random_state= 5)

In [511]:
lin_model= LinearRegression().fit(x_train, y_train)

In [512]:
yHat__= lin_model.predict(x_test)

In [513]:
print('RSMLE:',rmsle(np.absolute(yHat__), np.absolute(y_test) ))

RSMLE: 0.1691990155491588


In [517]:
X_= df_[:].values

In [518]:
for i in range(len(X_[:])):
    for j in range(len(X_[0])):
        arr= X_[i][j]
        if math.isnan(arr) :
            X_[i][j]= 0

In [519]:
x_train_, x_test_, y_train_, y_test_= train_test_split(X_, Y_Feature_, test_size= 0.3, random_state= 5)

In [520]:
modelALL= LinearRegression().fit(x_train_, y_train_)

In [523]:
yHatA= modelALL.predict(x_test_)
print('RSMLE:',rmsle(np.absolute(yHatA), np.absolute(y_test_) ))

RSMLE: 1.090272679065468e-15


In [525]:
testdf= pd.read_csv('C:/Users/rgangurde/Downloads/test.csv')

In [526]:
testdf.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0         Lvl    AllPub  ...         120        0    NaN  MnPrv         NaN   
1         Lvl    AllPub  ...           0        0    NaN    NaN        Gar2   
2         Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   
3         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
4         HLS    AllPub  ...         144        0    NaN    NaN         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0      6    2010        WD         Normal  
1   12500      6    2010        WD         Normal  
2       0      3    2010        WD         Normal  
3       0      6    2010        WD         Normal  
4       0      1    2010        WD         Normal  

[5 rows x 80 columns]